
<img src="https://raw.githubusercontent.com/rhatiro/Curso_EBAC-Profissao_Cientista_de_Dados/main/ebac-course-utils/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Profissão: Cientista de Dados**
### **Módulo 13** | Regressão II | Exercício 1

Aluno [Marcello Lassalla](https://www.linkedin.com/in/marcello-lassalla-a146b8225/)<br>
Data: 27 de junho de 2024.

---

# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda

Vamos trabalhar com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [1]:
import numpy as np
import pandas as pd

import patsy
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [2]:
filepath = 'https://raw.githubusercontent.com/MarcelloLM/Cienciadedados/main/Modulo%2010/previsao_de_renda.csv'
df = pd.read_csv(filepath_or_buffer=filepath)
df

Unnamed: 0    data_ref  index sexo  posse_de_veiculo  posse_de_imovel  \
0               0  2015-01-01   8682    F             False             True   
1               1  2015-01-01  12830    M              True             True   
2               2  2015-01-01   2885    M              True             True   
3               3  2015-01-01  16168    F              True            False   
4               4  2015-01-01  12438    M             False            False   
...           ...         ...    ...  ...               ...              ...   
14995       14995  2016-03-01   3678    M              True             True   
14996       14996  2016-03-01   3405    M              True             True   
14997       14997  2016-03-01   7944    F              True             True   
14998       14998  2016-03-01  16019    F             False             True   
14999       14999  2016-03-01  12895    F              True             True   

       qtd_filhos        tipo_renda           educacao estado_civil  \
0               1       Assalariado         Secundário       Casado   
1               0        Empresário         Secundário       Casado   
2               2       Assalariado  Superior completo       Casado   
3               0        Empresário         Secundário       Casado   
4               0       Assalariado         Secundário       Casado   
...           ...               ...                ...          ...   
14995           1       Assalariado         Secundário       Casado   
14996           1  Servidor público  Superior completo       Casado   
14997           0       Assalariado         Secundário       Casado   
14998           0  Servidor público  Superior completo       Casado   
14999           0       Assalariado         Secundário       Casado   

      tipo_residencia  idade  tempo_emprego  qt_pessoas_residencia    mau  \
0                Casa     36       3.575342                    3.0  False   
1                Casa     42       0.860274                    2.0  False   
2                Casa     31       8.065753                    4.0  False   
3                Casa     50       1.208219                    2.0  False   
4                Casa     52      13.873973                    2.0  False   
...               ...    ...            ...                    ...    ...   
14995            Casa     37       2.046575                    3.0  False   
14996         Estúdio     30      12.079452                    3.0  False   
14997            Casa     40       2.161644                    2.0  False   
14998            Casa     47       3.295890                    2.0  False   
14999            Casa     50       3.049315                    2.0  False   

         renda  
0      3369.24  
1      6096.14  
2      5658.98  
3      7246.69  
4      4017.37  
...        ...  
14995  1193.29  
14996  2519.18  
14997  1586.85  
14998  6975.39  
14999   993.86  

[15000 rows x 16 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             15000 non-null  int64  
 1   data_ref               15000 non-null  object 
 2   index                  15000 non-null  int64  
 3   sexo                   15000 non-null  object 
 4   posse_de_veiculo       15000 non-null  bool   
 5   posse_de_imovel        15000 non-null  bool   
 6   qtd_filhos             15000 non-null  int64  
 7   tipo_renda             15000 non-null  object 
 8   educacao               15000 non-null  object 
 9   estado_civil           15000 non-null  object 
 10  tipo_residencia        15000 non-null  object 
 11  idade                  15000 non-null  int64  
 12  tempo_emprego          12466 non-null  float64
 13  qt_pessoas_residencia  15000 non-null  float64
 14  mau                    15000 non-null  bool   
 15  re

1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
    - Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
    - Mantenha sempre a categoria mais frequente como casela de referência
    - Avalie os parâmetros e veja se parecem fazer sentido prático.  


2. Remova a variável menos significante e analise:
    - Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
    - Observe os parâmetros e veja se algum se alterou muito.  


3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.
    

---

#### 1. Ajuste um modelo para prever log(renda) considerando todas as covariáveis disponíveis.
- Utilizando os recursos do Patsy, coloque as variáveis qualitativas como *dummies*.
- Mantenha sempre a categoria mais frequente como casela de referência
- Avalie os parâmetros e veja se parecem fazer sentido prático.

In [4]:
# Analisando a frequência das categorias em cada variável:

for coluna in df.columns[3:]:
    print(coluna, '\n', df[coluna].value_counts(), '\n')

sexo 
 sexo
F    10167
M     4833
Name: count, dtype: int64 

posse_de_veiculo 
 posse_de_veiculo
False    9167
True     5833
Name: count, dtype: int64 

posse_de_imovel 
 posse_de_imovel
True     10082
False     4918
Name: count, dtype: int64 

qtd_filhos 
 qtd_filhos
0     10358
1      2992
2      1458
3       165
4        19
14        4
7         2
5         2
Name: count, dtype: int64 

tipo_renda 
 tipo_renda
Assalariado         7826
Empresário          3311
Pensionista         2543
Servidor público    1316
Bolsista               4
Name: count, dtype: int64 

educacao 
 educacao
Secundário             8927
Superior completo      5262
Superior incompleto     599
Primário                193
Pós graduação            19
Name: count, dtype: int64 

estado_civil 
 estado_civil
Casado      10550
Solteiro     1767
União        1168
Separado      838
Viúvo         677
Name: count, dtype: int64 

tipo_residencia 
 tipo_residencia
Casa             13566
Com os pais        637
Governamental  

In [5]:
y, X = patsy.dmatrices(formula_like='''
                                    np.log(renda) ~ sexo
                                                    + posse_de_veiculo
                                                    + posse_de_imovel
                                                    + C(qtd_filhos, Treatment(0))
                                                    + C(tipo_renda, Treatment('Assalariado'))
                                                    + C(educacao, Treatment('Secundário'))
                                                    + C(estado_civil, Treatment('Casado'))
                                                    + C(tipo_residencia, Treatment('Casa'))
                                                    + idade
                                                    + tempo_emprego
                                                    + C(qt_pessoas_residencia, Treatment(2))
                                    ''',
                       data=df)

In [6]:
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.243
Model:                            OLS   Adj. R-squared:                  0.241
Method:                 Least Squares   F-statistic:                     117.3
Date:                Thu, 27 Jun 2024   Prob (F-statistic):               0.00
Time:                        17:03:06   Log-Likelihood:                -13195.
No. Observations:               12466   AIC:                         2.646e+04
Df Residuals:                   12431   BIC:                         2.672e+04
Df Model:                          34                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.2562      0.038    192.707      0.000       7.182       7.330
sexo[T.M]                                                       0.4807      0.014     33.586      0.000       0.453       0.509
posse_de_veiculo[T.True]                                       -0.0014      0.014     -0.100      0.920      -0.028       0.026
posse_de_imovel[T.True]                                         0.0861      0.014      6.353      0.000       0.060       0.113
C(qtd_filhos, Treatment(0))[T.1]                               -0.0345      0.131     -0.264      0.792      -0.290       0.222
C(qtd_filhos, Treatment(0))[T.2]                                0.1056      0.267      0.395      0.693      -0.418       0.629
C(qtd_filhos, Treatment(0))[T.3]                               -0.4513      0.471     -0.958      0.338      -1.375       0.472
C(qtd_filhos, Treatment(0))[T.4]                               -0.1671      0.414     -0.403      0.687      -0.979       0.645
C(qtd_filhos, Treatment(0))[T.5]                               -0.0311      0.399     -0.078      0.938      -0.812       0.750
C(qtd_filhos, Treatment(0))[T.7]                               -0.4790      0.247     -1.939      0.053      -0.963       0.005
C(qtd_filhos, Treatment(0))[T.14]                               0.0646      0.187      0.346      0.729      -0.301       0.430
C(tipo_renda, Treatment('Assalariado'))[T.Bolsista]            -0.0490      0.350     -0.140      0.889      -0.734       0.636
C(tipo_renda, Treatment('Assalariado'))[T.Empresário]           0.1789      0.015     12.232      0.000       0.150       0.208
C(tipo_renda, Treatment('Assalariado'))[T.Pensionista]          0.3768      0.233      1.616      0.106      -0.080       0.834
C(tipo_renda, Treatment('Assalariado'))[T.Servidor público]     0.0936      0.021      4.420      0.000       0.052       0.135
C(educacao, Treatment('Secundário'))[T.Primário]               -0.0212      0.071     -0.298      0.765      -0.160       0.118
C(educacao, Treatment('Secundário'))[T.Pós graduação]           0.5503      0.161      3.418      0.001       0.235       0.866
C(educacao, Treatment('Secundário'))[T.Superior completo]       0.0962      0.013      7.145      0.000       0.070       0.123
C(educacao, Treatment('Secundário'))[T.Superior incompleto]    -0.0284      0.031     -0.917      0.359      -0.089       0.032
C(estado_civil, Treatment('Casado'))[T.Separado]               -0.0384      0.131     -0.293      0.769      -0.295       0.218
C(estado_civil, Treatment('Casado'))[T.Solteiro]                0.0209      0.129      0.162      0.871      -0.232       0.273
C(estado_ci

#### 2. Remova a variável menos significante e analise:
- Observe os indicadores que vimos, e avalie se o modelo melhorou ou piorou na sua opinião.
- Observe os parâmetros e veja se algum se alterou muito.

In [7]:
# Removendo a variável `tipo_residencia` com o menor P>|t| == 0.200:

smf.ols(formula='''
                np.log(renda) ~ sexo
                                + posse_de_veiculo
                                + posse_de_imovel
                                + C(qtd_filhos, Treatment(0))
                                + C(tipo_renda, Treatment('Assalariado'))
                                + C(educacao, Treatment('Secundário'))
                                + C(estado_civil, Treatment('Casado'))
                                + idade
                                + tempo_emprego
                                + C(qt_pessoas_residencia, Treatment(2))
                ''',
        data=df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.242
Model:                            OLS   Adj. R-squared:                  0.240
Method:                 Least Squares   F-statistic:                     137.0
Date:                Thu, 27 Jun 2024   Prob (F-statistic):               0.00
Time:                        17:03:07   Log-Likelihood:                -13201.
No. Observations:               12466   AIC:                         2.646e+04
Df Residuals:                   12436   BIC:                         2.668e+04
Df Model:                          29                                         
Covariance Type:            nonrobust                                         
===============================================================================================================================
                                                                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------------------------------
Intercept                                                       7.2372      0.037    196.640      0.000       7.165       7.309
sexo[T.M]                                                       0.4796      0.014     33.608      0.000       0.452       0.508
posse_de_veiculo[T.True]                                       -0.0018      0.014     -0.131      0.896      -0.029       0.025
posse_de_imovel[T.True]                                         0.0945      0.013      7.122      0.000       0.068       0.120
C(qtd_filhos, Treatment(0))[T.1]                               -0.0352      0.131     -0.269      0.788      -0.291       0.221
C(qtd_filhos, Treatment(0))[T.2]                                0.1021      0.267      0.382      0.703      -0.422       0.626
C(qtd_filhos, Treatment(0))[T.3]                               -0.4917      0.470     -1.045      0.296      -1.414       0.431
C(qtd_filhos, Treatment(0))[T.4]                               -0.1478      0.414     -0.357      0.721      -0.959       0.664
C(qtd_filhos, Treatment(0))[T.5]                               -0.0166      0.398     -0.042      0.967      -0.797       0.764
C(qtd_filhos, Treatment(0))[T.7]                               -0.4736      0.247     -1.916      0.055      -0.958       0.011
C(qtd_filhos, Treatment(0))[T.14]                               0.0664      0.187      0.356      0.722      -0.299       0.432
C(tipo_renda, Treatment('Assalariado'))[T.Bolsista]            -0.0499      0.350     -0.143      0.887      -0.735       0.636
C(tipo_renda, Treatment('Assalariado'))[T.Empresário]           0.1790      0.015     12.241      0.000       0.150       0.208
C(tipo_renda, Treatment('Assalariado'))[T.Pensionista]          0.3839      0.233      1.646      0.100      -0.073       0.841
C(tipo_renda, Treatment('Assalariado'))[T.Servidor público]     0.0908      0.021      4.298      0.000       0.049       0.132
C(educacao, Treatment('Secundário'))[T.Primário]               -0.0239      0.071     -0.337      0.736      -0.163       0.115
C(educacao, Treatment('Secundário'))[T.Pós graduação]           0.5550      0.161      3.447      0.001       0.239       0.871
C(educacao, Treatment('Secundário'))[T.Superior completo]       0.0966      0.013      7.181      0.000       0.070       0.123
C(educacao, Treatment('Secundário'))[T.Superior incompleto]    -0.0287      0.031     -0.930      0.353      -0.089       0.032
C(estado_civil, Treatment('Casado'))[T.Separado]               -0.0402      0.131     -0.307      0.759      -0.297       0.216
C(estado_civil, Treatment('Casado'))[T.Solteiro]                0.0176      0.129      0.136      0.891      -0.235       0.270
C(estado_ci

**Resposta:**

Após remover a variável `tipo_residencia`, o modelo não apresentou melhora nem piora significativa. O coeficiente de determinação manteve o mesmo valor de 0.358 e o *p-value* das variáveis teve uma variação muito pequena, praticamente insignificante.

#### 3. Siga removendo as variáveis menos significantes, sempre que o *p-value* for menor que 5%. Compare o modelo final com o inicial. Observe os indicadores e conclua se o modelo parece melhor.

In [8]:
smf.ols(formula='''
                np.log(renda) ~ sexo
                                + posse_de_veiculo
                                + posse_de_imovel
                                + tipo_renda == 'Empresário'
                                + tipo_renda == 'Servidor público'
                                + educacao == 'Superior completo'
                                + qt_pessoas_residencia == 3
                                + qt_pessoas_residencia == 4
                                + idade
                                + tempo_emprego
                ''',
        data=df).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.240
Model:                            OLS   Adj. R-squared:                  0.239
Method:                 Least Squares   F-statistic:                     393.3
Date:                Thu, 27 Jun 2024   Prob (F-statistic):               0.00
Time:                        17:03:08   Log-Likelihood:                -13219.
No. Observations:               12466   AIC:                         2.646e+04
Df Residuals:                   12455   BIC:                         2.654e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
============================================================================================================
                                               coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------
Intercept                                    7.2281      0.034    213.553      0.000       7.162       7.294
sexo[T.M]                                    0.4819      0.014     34.000      0.000       0.454       0.510
posse_de_veiculo[T.True]                     0.0018      0.014      0.133      0.894      -0.025       0.029
posse_de_imovel[T.True]                      0.0965      0.013      7.298      0.000       0.071       0.122
tipo_renda == 'Empresário'[T.True]           0.1792      0.015     12.290      0.000       0.151       0.208
tipo_renda == 'Servidor público'[T.True]     0.0914      0.021      4.324      0.000       0.050       0.133
educacao == 'Superior completo'[T.True]      0.0973      0.013      7.396      0.000       0.072       0.123
qt_pessoas_residencia == 3[T.True]           0.0056      0.017      0.341      0.733      -0.027       0.038
qt_pessoas_residencia == 4[T.True]           0.0572      0.021      2.740      0.006       0.016       0.098
idade                                        0.0053      0.001      7.080      0.000       0.004       0.007
tempo_emprego                                0.0482      0.001     48.464      0.000       0.046       0.050
==============================================================================
Omnibus:                        1.032   Durbin-Watson:                   2.015
Prob(Omnibus):                  0.597   Jarque-Bera (JB):                1.001
Skew:                          -0.012   Prob(JB):                        0.606
Kurtosis:                       3.037   Cond. No.                         241.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Resposta:**

Ao comparar o modelo final com o modelo inicial, os indicadores não apresentaram nenhuma melhora ou piora significativa. O coeficiente de determinação diminuiu de 0.358 para 0.357 e o AIC de 27190 para 27170.

---